datalad/support/protocol.py

```diff
@@ -0,0 +1,319 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+""" Protocolling  command calls.
+"""
+
+from abc import ABCMeta, abstractmethod, abstractproperty
+from os import linesep
+import logging
+import time
+
+lgr = logging.getLogger('datalad.protocol')
+
+
+class ProtocolInterface(object):
+    """Interface class for protocols used by the Runner.
+
+    Implementations of this interface are supposed to store one section per
+    recorded command call in `self._sections` as a dictionary. Default
+    implementation of __str__ has to be overridden otherwise.
+
+    ProtocolInterface is iterable as a list of sections.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self):
+        self._sections = []
+        self._title = ''
+
+    def __iter__(self):
+        return self._sections.__iter__()
+
+    def __getitem__(self, item):
+        return self._sections.__getitem__(item)
+
+    def __len__(self):
+        return len(self._sections)
+
+    def __str__(self):
+        protocol_str = self._title
+        for section in self._sections:
+            protocol_str += linesep
+            for key in section.keys():
+                protocol_str += "%s: %s" % (key, section[key]) + linesep
+
+        return protocol_str
+
+    @abstractmethod
+    def start_section(self, cmd):
+        """Starts a new section of the protocol.
+
+        To be used with a corresponding call of end_section().
+
+        Parameters:
+        -----------
+        cmd: list
+         The actual command and its options/arguments as a list
+
+        Returns:
+        --------
+        int
+          An id of the started section to be used as argument of the
+          corresponding call of end_section().
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def end_section(self, id_, exception):
+        """Ends the section `id`.
+
+        This ends the section defined by `id` as returned by start_section().
+
+        Parameters:
+        -----------
+        id: int
+        exception: Exception
+          The exception raised by the command if any or None otherwise.
+
+        Raises:
+        -------
+        IndexError
+          in case `id` is invalid.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_section(self, cmd, exception):
+        """Adds a section to the protocol.
+
+        This is an alternative to the use of start_section() and end_section().
```

Comment:
Well, depends on implementation whether or not there is a difference. It's supposed to provide the information, that it was called right before the call to be protocolled, whereas add_section doesn't tell you anything about that. So it's on the implementation of a certain protocol to use that information or not. ExecutionTimeProtocol makes use of it. Others don't.